File: R/tests/testthat/test-io-csv.R

# Helper — write a temp CSV and return its path
.write_tmp_csv <- function(df) {
  path <- tempfile(fileext = ".csv")
  utils::write.csv(df, path, row.names = FALSE)
  path
}

test_that("read_csv_eventlog reads minimal CSV with default column names", {
  df <- data.frame(
    case_id   = c("C1", "C1", "C2"),
    activity  = c("A", "B", "A"),
    timestamp = c("2024-01-01 00:00:00", "2024-01-01 01:00:00", "2024-01-01 00:00:00"),
    stringsAsFactors = FALSE
  )
  path <- .write_tmp_csv(df)
  log  <- read_csv_eventlog(path, timestamp_format = "%Y-%m-%d %H:%M:%S")

  expect_equal(nrow(log), 3L)
  expect_equal(log$case_id,  c("C1", "C1", "C2"))
  expect_equal(log$activity, c("A",  "B",  "A"))
  expect_s3_class(log$timestamp, "POSIXct")
  expect_equal(attr(log$timestamp, "tzone"), "UTC")
})

test_that("read_csv_eventlog maps custom column names", {
  df <- data.frame(
    CaseID    = c("X1", "X1"),
    EventName = c("Start", "End"),
    EventTime = c("2024-01-01 08:00:00", "2024-01-01 09:00:00"),
    stringsAsFactors = FALSE
  )
  path <- .write_tmp_csv(df)
  log  <- read_csv_eventlog(
    path,
    case_col      = "CaseID",
    activity_col  = "EventName",
    timestamp_col = "EventTime",
    timestamp_format = "%Y-%m-%d %H:%M:%S"
  )

  expect_equal(names(log)[1:3], c("case_id", "activity", "timestamp"))
  expect_equal(log$case_id, c("X1", "X1"))
})

test_that("read_csv_eventlog includes optional columns when specified", {
  df <- data.frame(
    case_id   = "C1",
    activity  = "A",
    timestamp = "2024-01-01 00:00:00",
    resource  = "alice",
    lifecycle = "complete",
    stringsAsFactors = FALSE
  )
  path <- .write_tmp_csv(df)
  log  <- read_csv_eventlog(
    path,
    resource_col  = "resource",
    lifecycle_col = "lifecycle",
    timestamp_format = "%Y-%m-%d %H:%M:%S"
  )

  expect_true("resource"  %in% names(log))
  expect_true("lifecycle" %in% names(log))
  expect_equal(log$resource,  "alice")
  expect_equal(log$lifecycle, "complete")
})

test_that("read_csv_eventlog folds columns into case_attrs and event_attrs", {
  df <- data.frame(
    case_id   = c("C1", "C1"),
    activity  = c("A",  "B"),
    timestamp = c("2024-01-01 00:00:00", "2024-01-01 01:00:00"),
    region    = c("NZ",  "NZ"),
    cost      = c("10",  "20"),
    stringsAsFactors = FALSE
  )
  path <- .write_tmp_csv(df)
  log  <- read_csv_eventlog(
    path,
    case_attrs_cols  = "region",
    event_attrs_cols = "cost",
    timestamp_format = "%Y-%m-%d %H:%M:%S"
  )

  expect_true("case_attrs"  %in% names(log))
  expect_true("event_attrs" %in% names(log))
  expect_equal(log$case_attrs[[1]][["region"]],  "NZ")
  expect_equal(log$event_attrs[[1]][["cost"]], "10")
  expect_equal(log$event_attrs[[2]][["cost"]], "20")
})

test_that("read_csv_eventlog converts non-UTC source timezone to UTC", {
  df <- data.frame(
    case_id   = "C1",
    activity  = "A",
    timestamp = "2024-01-01 12:00:00",
    stringsAsFactors = FALSE
  )
  path <- .write_tmp_csv(df)
  log  <- read_csv_eventlog(
    path,
    timestamp_format = "%Y-%m-%d %H:%M:%S",
    tz = "Pacific/Auckland"  # UTC+13
  )

  # 2024-01-01 12:00 NZDT = 2024-12-31 23:00 UTC
  expect_equal(attr(log$timestamp, "tzone"), "UTC")
})

test_that("read_csv_eventlog errors when required source column is missing", {
  df <- data.frame(
    case_id   = "C1",
    activity  = "A",
    timestamp = "2024-01-01 00:00:00",
    stringsAsFactors = FALSE
  )
  path <- .write_tmp_csv(df)
  expect_error(read_csv_eventlog(path, case_col = "no_such_col"), "not found in CSV")
})

test_that("read_csv_eventlog errors when optional source column is missing", {
  df <- data.frame(
    case_id   = "C1",
    activity  = "A",
    timestamp = "2024-01-01 00:00:00",
    stringsAsFactors = FALSE
  )
  path <- .write_tmp_csv(df)
  expect_error(
    read_csv_eventlog(path, resource_col = "no_resource",
                      timestamp_format = "%Y-%m-%d %H:%M:%S"),
    "not found in CSV"
  )
})

File: python/tests/test_io_csv.py
"""Tests for read_csv_eventlog."""

from __future__ import annotations

import textwrap

import pytest

from processmine_ml.io import read_csv_eventlog


def _csv_path(tmp_path, content: str):
    p = tmp_path / "log.csv"
    p.write_text(textwrap.dedent(content).strip())
    return p


def test_reads_minimal_csv_default_columns(tmp_path):
    p = _csv_path(tmp_path, """
        case_id,activity,timestamp
        C1,A,2024-01-01 00:00:00
        C1,B,2024-01-01 01:00:00
        C2,A,2024-01-01 00:00:00
    """)
    log = read_csv_eventlog(p)
    assert len(log) == 3
    assert list(log.columns[:3]) == ["case_id", "activity", "timestamp"]
    assert str(log["timestamp"].dtype.tz) == "UTC"


def test_maps_custom_column_names(tmp_path):
    p = _csv_path(tmp_path, """
        CaseID,EventName,EventTime
        X1,Start,2024-01-01 08:00:00
        X1,End,2024-01-01 09:00:00
    """)
    log = read_csv_eventlog(
        p,
        case_col="CaseID",
        activity_col="EventName",
        timestamp_col="EventTime",
    )
    assert list(log.columns[:3]) == ["case_id", "activity", "timestamp"]
    assert list(log["case_id"]) == ["X1", "X1"]


def test_includes_optional_columns(tmp_path):
    p = _csv_path(tmp_path, """
        case_id,activity,timestamp,resource,lifecycle
        C1,A,2024-01-01 00:00:00,alice,complete
    """)
    log = read_csv_eventlog(p, resource_col="resource", lifecycle_col="lifecycle")
    assert "resource" in log.columns
    assert "lifecycle" in log.columns
    assert log["resource"].iloc[0] == "alice"
    assert log["lifecycle"].iloc[0] == "complete"


def test_folds_columns_into_attrs(tmp_path):
    p = _csv_path(tmp_path, """
        case_id,activity,timestamp,region,cost
        C1,A,2024-01-01 00:00:00,NZ,10
        C1,B,2024-01-01 01:00:00,NZ,20
    """)
    log = read_csv_eventlog(
        p,
        case_attrs_cols=["region"],
        event_attrs_cols=["cost"],
    )
    assert "case_attrs" in log.columns
    assert "event_attrs" in log.columns
    assert log["case_attrs"].iloc[0]["region"] == "NZ"
    assert log["event_attrs"].iloc[0]["cost"] == "10"
    assert log["event_attrs"].iloc[1]["cost"] == "20"


def test_converts_non_utc_source_tz_to_utc(tmp_path):
    p = _csv_path(tmp_path, """
        case_id,activity,timestamp
        C1,A,2024-01-01 12:00:00
    """)
    log = read_csv_eventlog(p, tz="America/New_York")
    assert str(log["timestamp"].dtype.tz) == "UTC"
    # 12:00 EST = 17:00 UTC
    assert log["timestamp"].iloc[0].hour == 17


def test_explicit_timestamp_format(tmp_path):
    p = _csv_path(tmp_path, """
        case_id,activity,timestamp
        C1,A,01/15/2024 08:30
    """)
    log = read_csv_eventlog(p, timestamp_format="%m/%d/%Y %H:%M")
    assert log["timestamp"].iloc[0].month == 1
    assert log["timestamp"].iloc[0].day == 15


def test_errors_on_missing_required_column(tmp_path):
    p = _csv_path(tmp_path, """
        case_id,activity,timestamp
        C1,A,2024-01-01
    """)
    with pytest.raises(ValueError, match="not found in CSV"):
        read_csv_eventlog(p, case_col="no_such_col")


def test_errors_on_missing_optional_column(tmp_path):
    p = _csv_path(tmp_path, """
        case_id,activity,timestamp
        C1,A,2024-01-01
    """)
    with pytest.raises(ValueError, match="not found in CSV"):
        read_csv_eventlog(p, resource_col="no_resource")


def test_result_passes_validate_eventlog(tmp_path):
    p = _csv_path(tmp_path, """
        case_id,activity,timestamp
        C1,A,2024-01-01 00:00:00
        C1,B,2024-01-01 01:00:00
    """)
    from processmine_ml.io import validate_eventlog
    log = read_csv_eventlog(p)
    validate_eventlog(log)  # should not raise


def test_passes_extra_kwargs_to_read_csv(tmp_path):
    p = _csv_path(tmp_path, """
        case_id;activity;timestamp
        C1;A;2024-01-01 00:00:00
    """)
    log = read_csv_eventlog(p, sep=";")
    assert len(log) == 1
